controller/recurring_job_controller.go

```diff
@@ -220,6 +233,78 @@ func (control *RecurringJobController) syncRecurringJob(key string) (err error)
 	return nil
 }
 
+func (control *RecurringJobController) cleanupVolumeRecurringJob(recurringJob *longhorn.RecurringJob) error {
+	// Check if each group of the recurring job contains other recurring jobs.
+	// If No, it means the recurring job is the last job of the group then
+	// Longhorn will clean up this group labels for all volumes.
+	checkRecurringJobs, err := control.ds.ListRecurringJobs()
+	if err != nil {
+		return err
+	}
+	rmGroups := []string{}
+	for _, group := range recurringJob.Spec.Groups {
+		inUse := false
+		for _, checkRecurringJob := range checkRecurringJobs {
+			if checkRecurringJob.Name == recurringJob.Name {
+				continue
+			}
+			if util.Contains(checkRecurringJob.Spec.Groups, group) {
+				inUse = true
+				break
+			}
+		}
+		if !inUse {
+			rmGroups = append(rmGroups, group)
+		}
+	}
+
+	// delete volume labels
+	volumes, err := control.ds.ListVolumes()
+	if err != nil {
+		return err
+	}
+	for _, vol := range volumes {
+		jobs := datastore.MarshalLabelToVolumeRecurringJob(vol.Labels)
+		for jobName, job := range jobs {
+			if job.IsGroup {
+				if !util.Contains(rmGroups, jobName) {
+					continue
+				}
+				control.logger.Debugf("Clean up volume recurring job-group %v for %v", jobName, vol.Name)
+				if _, err := control.ds.DeleteVolumeRecurringJob(jobName, true, vol); err != nil {
+					return err
+				}
+			} else if jobName == recurringJob.Name {
+				control.logger.Debugf("Clean up volume recurring job %v for %v", jobName, vol.Name)
+				if _, err := control.ds.DeleteVolumeRecurringJob(jobName, false, vol); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (control *RecurringJobController) detachVolumeAutoAttachedByRecurringJob(name string, v *longhorn.Volume) error {
+	if v.Spec.LastAttachedBy != name {
+		return nil
+	}
+	// !volume.DisableFrontend condition makes sure that volume is detached by this recurring job,
```

Comment:
Why not just reuse `VolumeManager::Detach`? We should better leverage existing functions instead of updating the spec directly if the operation is general.

Reply:
The logic is similar to [detachShareManagerVolume](https://github.com/longhorn/longhorn-manager/blob/master/controller/share_manager_controller.go#L399). I don't think we can call VolumeManager in the controller? Or do you think we should new a VolumeManager for this?

Reply:
hmm, right now the volume manager seems the interface for the API instead. It's fine for now, but it looks like the code reuse can be improved a bit.